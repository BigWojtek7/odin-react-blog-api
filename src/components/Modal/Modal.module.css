.modal{
  margin: 100px auto;
  padding: 2em;;
  background-color: var(--nav-bg);
  color: var(--font-clr);
  border: var(--btn-clr) 5px solid;
}

.buttons{
  display: flex;
  justify-content: center;
  margin-top: 1.5em;
}

.buttons button{
  flex: 1
}

.modal::backdrop {
  background: hsl(0 0% 0% / 70%);
}

.children {
  margin-bottom: 0.5em;
}