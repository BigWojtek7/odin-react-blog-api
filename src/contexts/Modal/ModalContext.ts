import { createContext, MouseEventHandler } from 'react';

interface ModalDataType {
  message: string;
  onConfirm: MouseEventHandler<HTMLButtonElement>;
}

export interface ModalContextType {
  modalData: ModalDataType | null;
  openModal: (message: string, onConfirm: any) => void;
  closeModal: () => void;
}

const ModalContext = createContext<ModalContextType | null>(null);

export default ModalContext;
