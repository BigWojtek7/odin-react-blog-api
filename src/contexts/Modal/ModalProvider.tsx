import { MouseEventHandler, useState } from 'react';
import ModalContext from './ModalContext';
import { ChildrenProps } from '../../types/SharedInterfaces';

interface ModalDataType {
  message: string;
  onConfirm: MouseEventHandler<HTMLButtonElement>;
}

const ModalProvider = ({ children }: ChildrenProps) => {
  const [modalData, setModalData] = useState<ModalDataType | null>(null);

  const openModal = (
    message: string,
    onConfirm: MouseEventHandler<HTMLButtonElement>
  ) => {
    setModalData({ message, onConfirm });
  };
  const closeModal = () => {
    setModalData(null);
  };

  return (
    <ModalContext.Provider value={{ modalData, openModal, closeModal }}>
      {children}
    </ModalContext.Provider>
  );
};

export default ModalProvider;
