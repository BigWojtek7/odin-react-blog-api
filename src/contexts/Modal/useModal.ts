import { useContext } from 'react';
import ModalContext from './ModalContext'

const useModal = () => {
  const modalContext = useContext(ModalContext);

  if (!modalContext) {
    throw new Error(
      'Please use useModal inside the context of ModalProvider'
    );
  }
  return modalContext;
};

export default useModal;
