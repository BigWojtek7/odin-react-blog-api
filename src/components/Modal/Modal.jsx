import { createPortal } from 'react-dom';
import useModal from '../../hooks/useModal';

import styles from './Modal.module.css';

function Modal() {
  const { modalData, closeModal } = useModal();
  if (!modalData) return null;

  return (
    <>
      {createPortal(
        <div className={styles.overlay}>
          <div className={styles.content}>
            <p>{modalData.message}</p>
            <button onClick={modalData.onConfirm}>Yes</button>
            <button onClick={closeModal}>No</button>
          </div>
        </div>,
        document.getElementById('modal-root')
      )}
    </>
  );
}

export default Modal;
