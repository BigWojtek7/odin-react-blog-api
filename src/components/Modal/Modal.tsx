import { createPortal } from 'react-dom';
import useModal from '../../contexts/Modal/useModal';

import styles from './Modal.module.css';
import Button from '../form/Button/Button';

function Modal() {
  const { modalData, closeModal } = useModal();

  if (!modalData) return null;

  const modalRoot = document.getElementById('modal-root');

  return (
    <>
      {modalRoot &&
        createPortal(
          <div className={styles.overlay}>
            <div className={styles.content}>
              <p>{modalData.message}</p>
              <Button onClick={modalData.onConfirm}>Yes</Button>
              <Button onClick={closeModal} style={{ marginLeft: '0.5em' }}>
                No
              </Button>
            </div>
          </div>,
          modalRoot
        )}
    </>
  );
}

export default Modal;
