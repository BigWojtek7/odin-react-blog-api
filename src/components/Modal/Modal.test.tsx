import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Modal from './Modal';

// Inicjalizacja mockowanych funkcji
let mockOnConfirm = vi.fn();
let mockCloseModal = vi.fn();

vi.mock('react-dom', () => ({
  ...vi.importActual('react-dom'),
  createPortal: (element) => element,
}));

// Mockowanie hooka useModal
vi.mock('../../contexts/Modal/useModal.js', () => {
  return {
    default: () => ({
      modalData: {
        message: 'Are you sure you want to delete this post?',
        onConfirm: mockOnConfirm,
      },
      closeModal: mockCloseModal,
    }),
  };
});

describe('Modal component', () => {
  it('renders modal with message and buttons', () => {
    render(<Modal />);

    expect(
      screen.getByText('Are you sure you want to delete this post?')
    ).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /yes/i })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /no/i })).toBeInTheDocument();
  });

  it('calls onConfirm when "Yes" button is clicked', async () => {
    const user = userEvent.setup();
    render(<Modal />);

    const yesButton = screen.getByRole('button', { name: /yes/i });
    await user.click(yesButton);

    expect(mockOnConfirm).toHaveBeenCalled();
  });

  it('closes the modal when "No" button is clicked', async () => {
    const user = userEvent.setup();
    render(<Modal />);

    const noButton = screen.getByRole('button', { name: /no/i });
    await user.click(noButton);

    expect(mockCloseModal).toHaveBeenCalled();
  });
});
