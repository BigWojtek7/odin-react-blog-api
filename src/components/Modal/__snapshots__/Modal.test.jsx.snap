// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Modal component > renders Loader component 1`] = `<div />`;

exports[`renders Loader component 1`] = `<div />`;
